cmd/protoc-gen-jsonshim/jsonshim/plugin.go

```diff
@@ -67,14 +67,24 @@ func (p *Plugin) Generate(file *generator.FileDescriptor) {
 	unmarshalerName := generator.FileName(file) + "Unmarshaler"
 	for _, message := range file.Messages() {
 		// check to make sure something was generated for this type
-		if !gogoproto.HasTypeDecl(file.FileDescriptorProto, message.DescriptorProto) || len(message.GetOneofDecl()) == 0 {
+		if !gogoproto.HasTypeDecl(file.FileDescriptorProto, message.DescriptorProto) {
+			continue
+		}
+
+		// skip maps in protos.
+		if message.Options != nil && message.Options.GetMapEntry() {
+			continue
+		}
+
+		// skip non top-level message with no oneof fields.
+		if len(message.TypeName()) > 1 && len(message.GetOneofDecl()) == 0 {
```

Comment:
I thought the reason we were generating marshallers for all types was because the field name mappings were not correct (i.e. json tag not matching pb tag, but pb tag name is used when serialized as json).